src/BleOtaUploader.cpp

```diff
@@ -1,288 +1,499 @@
+        if (status != BleOtaStatus::Ok)
+        {
+            terminateUpload(status);
+            handleError(status);
+            return;
+        }
+    }
+#endif
+
+    _storage.close();
+    _state = State::Install;
+
+    sendMessage(BleOtaEndResp{});
+
+    _uploadCallbacks->handleUploadEnd();
+}
+
+void BleOtaUploader::handleSignatureReq(const BleOtaSignatureReq& req)
+{
+    BLE_OTA_LOG(TAG, "Handle SignatureReq");
+
+    if (_state != State::Upload)
+    {
+        handleError(_state == State::Terminate ? _terminateCode : BleOtaStatus::UploadStopped);
         return;
     }
 
-    #ifndef BLE_OTA_NO_BUFFER
-    flushBuffer();
-    #endif
+    if (not _signature.isEnabled())
+    {
+        terminateUpload(BleOtaStatus::SignatureNotSupported);
+        handleError(BleOtaStatus::SignatureNotSupported);
+        return;
+    }
 
-    send(OK);
+    const auto status = _signature.pushSignature(req.data, req.size);
+    if (status != BleOtaStatus::Ok)
+    {
+        terminateUpload(status);
+        handleError(status);
+        return;
+    }
 
-    ArduinoBleOTA.uploadCallbacks->onEnd();
-    installing = true;
+    sendMessage(BleOtaSignatureResp{});
 }
 
-void BleOtaUploader::handleSetPinCode(const uint8_t* data, size_t length)
+void BleOtaUploader::handleSetPinReq(const BleOtaSetPinReq& req)
 {
-    if (uploading)
+    BLE_OTA_LOG(TAG, "Handle SetPinReq: pin: %lu", req.pin);
+
+    if (_state == State::Upload)
     {
-        handleError(NOK);
+        handleError(BleOtaStatus::UploadRunning);
         return;
     }
-    if (length != sizeof(uint32_t))
+
+    if (_pinCallbacks == &dummyPinCallbacks)
     {
-        handleError(INCORRECT_FORMAT);
+        handleError(BleOtaStatus::PinNotSupported);
         return;
     }
 
-    uint32_t pinCode;
-    memcpy(&pinCode, data, length);
-    send(ArduinoBleOTA.securityCallbacks->setPinCode(pinCode) ? OK : NOK);
+    _pinCallbacks->setPinCode(req.pin) ?
+        sendMessage(BleOtaSetPinResp{}) :
+        handleError(BleOtaStatus::PinChangeError);
 }
 
-void BleOtaUploader::handleRemovePinCode(const uint8_t* data, size_t length)
+void BleOtaUploader::handleRemovePinReq(const BleOtaRemovePinReq& req)
 {
-    if (uploading)
+    BLE_OTA_LOG(TAG, "Handle RemovePinReq");
+
+    if (_state == State::Upload)
     {
-        handleError(NOK);
+        handleError(BleOtaStatus::UploadRunning);
         return;
     }
-    if (length)
+
+    if (_pinCallbacks == &dummyPinCallbacks)
     {
-        handleError(INCORRECT_FORMAT);
+        handleError(BleOtaStatus::PinNotSupported);
         return;
     }
 
-    send(ArduinoBleOTA.securityCallbacks->removePinCode() ? OK : NOK);
+    _pinCallbacks->removePinCode() ?
+        sendMessage(BleOtaRemovePinResp{}) :
+        handleError(BleOtaStatus::PinChangeError);
 }
 
 void BleOtaUploader::handleInstall()
 {
     delay(250);
-    storage->close();
-    delay(250);
-    storage->apply();
+    _storage.apply();
     while (true);
```

Comment:
?

Reply:
Or you mean why `handleInstall()` is needed. Better to restart device in main thread in `pull()` method when we finish all tasks in `loop()` and not terminate in the middle (for example in the middle of saving settings)